synapse/cores/sortedram.py

```diff
@@ -0,0 +1,222 @@
+import collections
+
+import sortedcontainers
+
+import synapse.cores.common
+from synapse.compat import isint
+
+
+class SortedRamCortex(synapse.cores.common.Cortex):
+    '''
+    SortedRamCortex is a RAM-backed Cortex that maintain int prop values
+     in sorted order to facilitate range queries.
+
+    Insertion performance appears to be a constant factor of 10% slower than
+     synapse.cores.ram.Cortex
+    Query by range performance is strictly better.
+    Query by range runtime is O(log n) vs O(n) for the ram:// cortex.
+    '''
+
+    def _initCortex(self):
+
+        def getValu(row):
+            return row[2]
+
+        def makePropSet():
+            return sortedcontainers.SortedSet(key=getValu)
+
+        self.rowsbyid = collections.defaultdict(set)
+        # rows by props with type(value) == string
+        self.srowsbyprop = collections.defaultdict(set)
+        # rows by props with type(value) == int
+        # split because sortedcontainers.SortedSet does not support keys with mixed types
+        self.irowsbyprop = collections.defaultdict(makePropSet)
+        self.rowsbyvalu = collections.defaultdict(set)
+
+        self.initSizeBy('ge',self._sizeByGe)
+        self.initRowsBy('ge',self._rowsByGe)
+
+        self.initSizeBy('le',self._sizeByLe)
+        self.initRowsBy('le',self._rowsByLe)
+
+        self.initSizeBy('range',self._sizeByRange)
+        self.initRowsBy('range',self._rowsByRange)
+
+    def _sizeByRange(self, prop, valu, limit=None):
+        min = None
+        if valu[0] is not None:
```

Comment:
please avoid the use of "is".  There are *so* many nasty edge cases. ( i've even seen a C extension that instantiated and returned it's own None such that None was not None. :)

Also this old (in this case unrelated but equally scairy) ditty:

```
>>> x = 10
>>> x is 10
True
>>> x = 100000000
>>> x is 100000000
False
```